### PR TITLE
Search and Home now have an implemetation in MVVM

### PR DESCRIPTION
**Pull request by JosephChing**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -92,3 +92,5 @@ iOSInjectionProject/
 # Netowkring configurations like API keys
 
 NetworkConfiguration.swift
+*.xcuserstate
+NetflixClone/Classes/Scenes/Home/.DS_Store
```
Comment (ahmdmhasn) [inline .gitignore]:
```suggestion
.DS_Store
```
It's enough to ignore any file with `.DS_Store` name

File: NetflixClone.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved [hunk 2]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (ahmdmhasn) [inline NetflixClone.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved]: I think we need to iterate over the files to remove files mistakenly staged to git and remove them. 

This thread is what I'm talking about https://stackoverflow.com/a/1274447/10798716. We need to run `git rm --cached .` but let's make it in a separate pull request to make changes small and trackable.

File: .swiftlint.yml [hunk 3]
```diff
@@ -0,0 +1,3 @@
+# By default, SwiftLint uses a set of sensible default rules you can adjust:
+disabled_rules: # rule identifiers turned on by default to exclude from running
+  - trailing_whitespace
```
Comment (ahmdmhasn) [inline .swiftlint.yml:3]: Great 🚀

File: NetflixClone/Classes/Models/MovieSearchMulti.swift [hunk 4]
```diff
@@ -0,0 +1,48 @@
+//
```
Comment (ahmdmhasn) [inline NetflixClone/Classes/Models/MovieSearchMulti.swift:1]: Let's use `Movie` object to avoid duplicating similar objects.

File: NetflixClone/Classes/Models/Movie.swift [hunk 5]
```diff
@@ -0,0 +1,43 @@
+//
+//  Movie.swift
+//  NetflixClone
+//
+//  Created by Joseph Ching on 2022-09-27.
+//
+
+import Foundation
+
+struct Movie: Codable, Hashable {
```
Comment (ahmdmhasn) [inline NetflixClone/Classes/Models/Movie.swift]:
```suggestion
struct Movie: Decodable, Hashable {
```
Since we will use it for decoding purposes only.

File: NetflixClone/Classes/Models/Movie.swift [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        case overview, popularity
+        case posterPath = "poster_path"
+        case releaseDate = "release_date"
+        case title, video
+        case voteAverage = "vote_average"
+        case voteCount = "vote_count"
+    }
+}
+
+struct Item: Sendable, Hashable {
```
Comment (ahmdmhasn) [inline NetflixClone/Classes/Models/Movie.swift:38]: Is it needed? if yes, we can add it as a nested type to movie
extension Movie { 
     struct Item {
          let name: String
          let idNumber: Int
          let rating: Double
          let tags: [String]
     }
}

File: NetflixClone/Classes/Networking/DiscoverMoviesRemote.swift [hunk 7]
```diff
@@ -8,14 +8,57 @@
...
+    enum MediaType: String {
+        case all
+        case movie
+        case tv
+        case person
+    }
+    enum TimeWindow: String {
+        case day
+        case week
+    }
```
Comment (ahmdmhasn) [inline NetflixClone/Classes/Networking/DiscoverMoviesRemote.swift]: Let's add enums in an extension to `DiscoverMoviesRemote` and organize code with `MARK: SOME SECTION`

File: NetflixClone/Classes/Networking/DiscoverMoviesRemote.swift [hunk 8]
```diff
@@ -8,14 +8,57 @@
...
+    func trendingMovies(at page: Int, type: MediaType, time: TimeWindow ) async throws -> [Movie] {
+        let urlString = "\(Config.baseURL)trending/\(type.rawValue)/\(time.rawValue)?api_key=\(Config.apiKey)&page=\(page)"
```
Comment (ahmdmhasn) [inline NetflixClone/Classes/Networking/DiscoverMoviesRemote.swift:32]: We can use URL components to construct the URL instead of the repetitive URLs 
More: https://www.swiftbysundell.com/articles/constructing-urls-in-swift/

File: NetflixClone/Classes/Networking/NetworkConfiguration.swift [hunk 9]
```diff
@@ -0,0 +1,18 @@
+//
+//  NetworkConfiguration.swift
+//  NetflixClone
+//
+//  Created by Joseph Ching on 2022-09-19.
+//
+
+import Foundation
+
+class NetworkConfiguration {
```
Comment (ahmdmhasn) [inline NetflixClone/Classes/Networking/NetworkConfiguration.swift:10]:
```suggestion
enum NetworkConfiguration {
```
It's common to use case-less enum for constants.

File: NetflixClone/Classes/Networking/PosterCollectionViewCell.swift [hunk 10]
```diff
@@ -0,0 +1,53 @@
+//
+//  PosterCollectionViewCell.swift
+//  NetflixClone
+//
+//  Created by Joseph Ching on 2022-09-19.
+//
+
+import UIKit
+class PosterCollectionViewCell: UICollectionViewCell {
```
Comment (ahmdmhasn) [inline NetflixClone/Classes/Networking/PosterCollectionViewCell.swift:9]: Let's use XIBs to make the design for minimal code.

File: NetflixClone/Classes/Scenes/Home/HomeViewController.swift [hunk 11]
```diff
@@ -6,125 +6,192 @@
...
-    enum Defaults {
-        static let moviesPerPage = 20
-    }
+class HomeViewController: UIViewController {
+    // MARK: Properties
+    typealias Section = HomeViewModel.Section
     
-    // MARK: Outlets
+    private var viewModel: HomeViewModel
+    private var subscribers: [AnyCancellable] = []
```
Comment (ahmdmhasn) [inline NetflixClone/Classes/Scenes/Home/HomeViewController.swift:16]:
```suggestion
    private var subscribers: Set<AnyCancellable> = []
```
It should be a Set, not a List; Since it helps random item selection.

File: NetflixClone/Classes/Scenes/Home/HomeViewController.swift [hunk 12]
```diff
@@ -6,125 +6,192 @@
...
+
+    var dataSource: UICollectionViewDiffableDataSource<Section, Movie>! = nil
+    var currentSnapshot: NSDiffableDataSourceSnapshot<Section, Movie>! = nil
```
Comment (ahmdmhasn) [inline NetflixClone/Classes/Scenes/Home/HomeViewController.swift]:
```suggestion
    var dataSource: UICollectionViewDiffableDataSource<Section, Movie>!
    var currentSnapshot: NSDiffableDataSourceSnapshot<Section, Movie>!
```
If you force unwrapping a property, no need to set it to `nil` initially.

File: NetflixClone/Classes/Scenes/Home/HomeViewModel.swift [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+
+class HomeViewModel {
+    
+    // MARK: - Properties
+    private let remote = DiscoverMoviesRemote()
+    private var hasMoreMovies = true
+    
+    var cancellables: Set<AnyCancellable> = []
+        
+    @Published var result: NewMovies? = NewMovies(newMovies: [], section: .featured)
```
Comment (ahmdmhasn) [inline NetflixClone/Classes/Scenes/Home/HomeViewModel.swift]:
```suggestion
        
    @Published private(set) var result = NewMovies(newMovies: [], section: .featured)
```

File: NetflixClone/Classes/Scenes/Home/HomeViewModel.swift [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+    
+    struct NewMovies {
```
Comment (ahmdmhasn) [inline NetflixClone/Classes/Scenes/Home/HomeViewModel.swift:21]: I like this ✅

File: NetflixClone/Classes/Scenes/Search/SearchViewController.swift [hunk 15]
```diff
@@ -6,43 +6,178 @@
...
+        configureDataSource()
         
-        let nibName = "\(SearchCollectionViewCell.self)"
-        searchCollectionView.register(UINib(nibName: nibName, bundle: nil), forCellWithReuseIdentifier: nibName)
+        viewModel.$result
+            .receive(on: DispatchQueue.main)
+            .sink { result in
+                if let result = result {
+                    self.updateDataSource(movies: result.newMovies)
+                }
```
Comment (ahmdmhasn) [inline NetflixClone/Classes/Scenes/Search/SearchViewController.swift]:
```suggestion
            .receive(on: DispatchQueue.main)
            .compactMap { $0 }
            .sink { [weak self] result in
                self?.updateDataSource(movies: result.newMovies)
            }
```

File: NetflixClone/Classes/Scenes/Search/SearchViewController.swift [hunk 16]
```diff
@@ -6,43 +6,178 @@
...
-    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
-        return 10
+// MARK: - Configurations
+extension SearchViewController {
+    private func configureHierarchy() {
+        stackView = UIStackView(frame: view.bounds)
+        stackView.axis = .vertical
+        stackView.alignment = .fill
+        searchBar = UISearchBar(frame: CGRect(x: 0.0, y: 0.0, width: view.frame.width, height: 48.0))
+        searchBar.delegate = self
```
Comment (ahmdmhasn) [inline NetflixClone/Classes/Scenes/Search/SearchViewController.swift:67]: Let's focus in making the layout in XIB files to avoid adding much code in view controller.

File: NetflixClone/Classes/Scenes/Home/HomeViewController.swift [hunk 17]
```diff
@@ -6,125 +6,192 @@
...
+                contentConfiguration.textProperties.color = UIColor.black
+                contentConfiguration.textProperties.font = UIFont.boldSystemFont(ofSize: 19)
+                supplementaryCell.contentConfiguration = contentConfiguration
+                supplementaryCell.backgroundConfiguration = .clear()
+            }
+        }
+    }
+    private func configureDataSource() {
+        dataSource = UICollectionViewDiffableDataSource
+        <Section, Movie>(collectionView: collectionView) {
```
Comment (ahmdmhasn) [inline NetflixClone/Classes/Scenes/Home/HomeViewController.swift]:
```suggestion
    typealias DataSource = UICollectionViewDiffableDataSource<Section, Movie>
    
    private func configureDataSource() {
        dataSource = DataSource(collectionView: collectionView) {
```
Typealiasss are great to minimize long repetitive names.

**Review by ahmdmhasn — COMMENTED** [review]
- Let's check SwiftFormat in the next session to apply auto-code formatting.
- We need to apply `.gitignore` to remove the files wrongly added to the repo. 

Great effort 🔥

File: .gitignore (hunk 1 above)
Comment (JosephChing) [reply .gitignore]: Thanks🔥! Noted for the future.

File: NetflixClone.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved (hunk 2 above)
Comment (JosephChing) [reply NetflixClone.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved]: Will do in a separate pull request! 👍

File: NetflixClone/Classes/Models/MovieSearchMulti.swift (hunk 4 above)
Comment (JosephChing) [reply NetflixClone/Classes/Models/MovieSearchMulti.swift:1]: Good idea! Thanks🔥.

File: NetflixClone/Classes/Scenes/Home/HomeViewModel.swift (hunk 13 above)
Comment (JosephChing) [reply NetflixClone/Classes/Scenes/Home/HomeViewModel.swift]: Will make sure that variables like this will be private(set) from now on 👍.

File: NetflixClone/Classes/Scenes/Home/HomeViewController.swift (hunk 17 above)
Comment (JosephChing) [reply NetflixClone/Classes/Scenes/Home/HomeViewController.swift]: I'll add typaliases for any other UICollectionViewDiffableDataSource from now on. Much cleaner🔥! Thanks!
